lib/db.js

```diff
@@ -453,6 +481,70 @@ module.exports = function (
     })
   }
 
+  DB.prototype.createDevice = function (uid, sessionTokenId, deviceInfo, noRetryOnConflict) {
+    log.trace({ op: 'DB.createDevice', uid: uid, id: deviceInfo.id })
+    var self = this
+    deviceInfo.id = crypto.randomBytes(16)
+    deviceInfo.createdAt = Date.now()
+    return this.pool.put(
+      '/account/' + uid.toString('hex') +
+      '/device/' + deviceInfo.id.toString('hex'),
+      unbuffer({
+        sessionTokenId: sessionTokenId,
+        createdAt: deviceInfo.createdAt,
+        name: deviceInfo.name,
+        type: deviceInfo.type,
+        callbackURL: deviceInfo.pushCallback,
+        callbackPublicKey: deviceInfo.pushPublicKey
+      })
+    )
+    .then(
+      function () {
+        return deviceInfo
+      },
+      function (err) {
+        if (isRecordAlreadyExistsError(err)) {
+          // It's possible (but extraordinarily improbable) that we generated
+          // a duplicate id. Retry once before failing with a session conflict.
```

Comment:
We could unambiguously determine the cause of the failure by fetching the existing device records and seeing where the conflict was.  If we're going to pay a second query to the db anyway, it may be worth making it unambiguous rather than just reducing the (already extraordinarily improbable) odds.